core/node/api_server/src/tx_sender/whitelisted_deploy_pool_sink.rs

```diff
@@ -0,0 +1,64 @@
+use zksync_dal::transactions_dal::L2TxSubmissionResult;
+use zksync_multivm::interface::{tracer::ValidationTraces, TransactionExecutionMetrics};
+use zksync_types::l2::L2Tx;
+
+use super::{allow_list_service::AllowListService, tx_sink::TxSink, SubmitTxError};
+use crate::tx_sender::master_pool_sink::MasterPoolSink;
+
+#[derive(Debug)]
+/// Wrapper that submits transactions to the mempool and enforces contract deployment allow-list.
+pub struct WhitelistedDeployPoolSink {
+    master_pool_sink: MasterPoolSink,
+    allowlist_service: AllowListService,
+}
+
+impl WhitelistedDeployPoolSink {
+    pub fn new(master_pool_sink: MasterPoolSink, allowlist_service: AllowListService) -> Self {
+        Self {
+            master_pool_sink,
+            allowlist_service,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl TxSink for WhitelistedDeployPoolSink {
+    async fn submit_tx(
+        &self,
+        tx: &L2Tx,
+        execution_metrics: TransactionExecutionMetrics,
+        validation_traces: ValidationTraces,
+    ) -> Result<L2TxSubmissionResult, SubmitTxError> {
+        let initiator = tx.initiator_account();
+
+        // Only enforce the allow-list check if VM actually deployed a contract.
+        if execution_metrics.vm.contract_deployment_count > 0 {
+            let initiator_allowed = self.allowlist_service.is_address_allowed(&initiator).await;
+
+            let contract_address_allowed = self
+                .allowlist_service
+                .is_address_allowed(&tx.execute.contract_address.unwrap_or_default())
```

Comment:
`unwrap_or_default()` looks hacky here; `contract_address == None` corresponds to EVM deployments. It'd be more straightforward to use `if let` here:

```rust
let contract_address_allowed = if let Some(addr) = &tx.execute.contract_address {
    self.allowlist_service.is_address_allowed(addr).await
} else {
    false // EVM deployments require the initiator to be whitelisted
};
```